File: kotlin-electron/karakum/nameResolvers/resolveInterfaceMethodNullableCallbackParameterName.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const callbackParameter = getParent(node)
    if (!callbackParameter) return null
    if (!ts.isParameter(callbackParameter)) return null
    if (!ts.isIdentifier(callbackParameter.name)) return null

    const callbackParameterName = callbackParameter.name.text

    const functionType = getParent(callbackParameter)
    if (!functionType) return null
    if (!ts.isFunctionTypeNode(functionType)) return null

    let union = getParent(functionType)
    if (ts.isParenthesizedTypeNode(union)) {
        union = getParent(union)
    }
    if (!union) return null
    if (!ts.isUnionTypeNode(union)) return null
    if (union.types.length < 2) return null

    let nullKeyword = union.types[1]
    if (ts.isParenthesizedTypeNode(nullKeyword)) {
        nullKeyword = nullKeyword.type
    }
    if (
        !ts.isLiteralTypeNode(nullKeyword)
        || nullKeyword.literal.kind !== ts.SyntaxKind.NullKeyword
    ) return null

    const parameter = getParent(union)
    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null

    const parameterName = parameter.name.text

    const method = getParent(parameter)
    if (!method) return null
    if (!ts.isMethodDeclaration(method)) return null
    if (!ts.isIdentifier(method.name)) return null

    const methodName = method.name.text

    const classNode = getParent(method)
    if (!classNode) return null
    if (!ts.isClassDeclaration(classNode)) return null
    if (classNode.name === undefined) return null

    const parentName = classNode.name.text

    return `${karakum.capitalize(parentName)}${karakum.capitalize(methodName)}${karakum.capitalize(parameterName)}${karakum.capitalize(callbackParameterName)}`
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/nameResolvers/resolveInterfaceMethodNullableCallbackParameterName.kt
package electron.karakum.nameResolvers

import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.Node
import typescript.SyntaxKind
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isFunctionTypeNode
import typescript.isIdentifier
import typescript.isLiteralTypeNode
import typescript.isMethodDeclaration
import typescript.isParameter
import typescript.isParenthesizedTypeNode
import typescript.isUnionTypeNode

fun resolveInterfaceMethodNullableCallbackParameterName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val callbackParameter = ensureNotNull(typeScriptService.getParent(node))
    ensure(isParameter(callbackParameter))

    val callbackParameterNameNode = callbackParameter.name
    ensure(isIdentifier(callbackParameterNameNode))
    val callbackParameterName = callbackParameterNameNode.text

    val functionType = ensureNotNull(typeScriptService.getParent(callbackParameter))
    ensure(isFunctionTypeNode(functionType))

    val union = ensureNotNull(
        typeScriptService.getParent(functionType)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isUnionTypeNode(union))
    ensure(union.types.asArray().size >= 2)

    val nullKeyword = union.types.asArray()[1].let {
        if (isParenthesizedTypeNode(it)) {
            it.type
        } else it
    }
    ensure(isLiteralTypeNode(nullKeyword))
    ensure(nullKeyword.literal.kind == SyntaxKind.NullKeyword)

    val parameter = ensureNotNull(typeScriptService.getParent(union))
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text

    val method = ensureNotNull(typeScriptService.getParent(parameter))
    ensure(isMethodDeclaration(method))

    val methodNameNode = method.name
    ensure(isIdentifier(methodNameNode))
    val methodName = methodNameNode.text

    val classNode = ensureNotNull(typeScriptService.getParent(method))
    ensure(isClassDeclaration(classNode))
    val parentName = ensureNotNull(classNode.name).text

    "${parentName.replaceFirstChar { it.titlecase() }}${methodName.replaceFirstChar { it.titlecase() }}${parameterName.replaceFirstChar { it.titlecase() }}${callbackParameterName.replaceFirstChar { it.titlecase() }}"
}
